--- variant.hpp.backup	2018-09-14 17:24:59.698450269 +0300
+++ variant.hpp	2018-09-14 17:25:22.208540255 +0300
@@ -789,7 +789,7 @@
 
   class bad_variant_access : public std::exception {
     public:
-    virtual const char *what() const noexcept { return "bad_variant_access"; }
+    virtual const char *what() const noexcept override { return "bad_variant_access"; }
   };
 
   [[noreturn]] inline void throw_bad_variant_access() {
